container_api/src/container.rs

```diff
@@ -124,45 +125,100 @@ impl Container {
-                }
-                Err(error) => Some(format!(
-                    "Error while trying to create instance \"{}\": {}",
-                    id, error
-                )),
-            })
-            .collect();
+        self.shutdown().map_err(|e| e.to_string())?;
+        self.instances = HashMap::new();
 
-        if errors.len() == 0 {
-            self.instances = instances;
-            Ok(())
-        } else {
-            Err(errors.iter().nth(0).unwrap().clone())
+        for id in config.instance_ids_sorted_by_bridge_dependencies()? {
+            let instance = self.instantiate_from_config(
+                &id,
+                &config,
+            )
+            .map_err(|error| format!(
+                "Error while trying to create instance \"{}\": {}",
+                id, error
+            ))?;
+
+            self.instances.insert(id.clone(), Arc::new(RwLock::new(instance)));
         }
+        Ok(())
     }
 
+    /// Creates one specific Holochain instance from a given Configuration,
+    /// id string and DnaLoader.
+    pub fn instantiate_from_config(
+        &mut self,
+        id: &String,
+        config: &Configuration,
+    ) -> Result<Holochain, String> {
+        let _ = config.check_consistency()?;
+
+        config
+            .instance_by_id(&id)
+            .ok_or(String::from("Instance not found in config"))
+            .and_then(|instance_config| {
+                // Build context:
+                let mut context_builder = ContextBuilder::new();
+
+                // Agent:
+                let agent_config = config.agent_by_id(&instance_config.agent).unwrap();
+                let pub_key = KeyBuffer::with_corrected(&agent_config.public_address)?;
+                context_builder.with_agent(AgentId::new(&agent_config.name, &pub_key));
+
+                // Network config:
+                instance_config.network.map(|network_config| {
+                    context_builder.with_network_config(JsonString::from(network_config))
+                });
+
+                // Storage:
+                if let StorageConfiguration::File { path } = instance_config.storage {
+                    context_builder
+                        .with_file_storage(path)
+                        .map_err(|hc_err| format!("Error creating context: {}", hc_err.to_string()))?;
+                }
+
+                // Container API
+                let mut api_builder = ContainerApiBuilder::new();
+                // Bridges:
+                let id = instance_config.id.clone();
+                for bridge in config.bridge_dependencies(id.clone()) {
+                    assert_eq!(bridge.caller_id, id.clone());
+                    let callee_config = config.instance_by_id(&bridge.callee_id)
+                        .expect("config.check_consistency()? jumps out if config is broken");
+                    let callee_instance = self.instances.get(&bridge.callee_id)
+                        .expect(r#"
```

Comment:
If we are returning a result, do we need these expects? same question of asserts

Reply:
If those expectations aren't met, it would not be a runtime error but a fundamental mistake in this approach. I assert that this can never happen with the current code. If it would we need to learn about it asap and fix it. If this panics because of changes to the code base I would want the developer to introduce changes to get drawn to this code and see what's going on.